### ISSUE TITLE
[BUG][Fuzzer][wrong-code] software pipelining silently reorders a read past an `atomic_add` to the same buffer, giving a stale result

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2595]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11 (source pinned to tag `v0.1.11`, commit `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`). Also live on `main` (see Additional context).

### System information

- TileLang `0.1.11` (source `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).
- Python 3.12, linux; PyTorch 2.12.1+cu130, CUDA 13.0; tilelang 0.1.11 (tag `v0.1.11` = `cd37ed5f`).
- GPU: NVIDIA H100 80GB HBM3 (compute capability 9.0, sm_90); Python 3.12, torch 2.12.1+cu130 (CUDA 13.0), tilelang 0.1.11 (tag `v0.1.11` = `cd37ed5f`). The defect is in a **target-independent TIR pass** (software-pipeline dependency analysis), so it is not arch-gated — reproduced on sm_90 this session; see the arch matrix in Additional context.

### Problem description

A loop-carried read-after-write dependency that runs **through an atomic** is invisible to the software pipeliner, so with `num_stages>=2` the dependent read is hoisted **ahead** of the `atomic_add` write to the same element and returns the stale (pre-atomic) value. The kernel compiles and runs to completion with no error or warning; the result is silently wrong and deterministic. `num_stages=1` (no pipeline) is correct, so pipelining alone changes the result.

In the repro below, iteration `ko` does `atomic_add(G[ko+1], 100)` and then reads `G[ko]` (the element the *previous* iteration bumped). With `num_stages=1` the reads see the bumped values (`[0,101,102,…,107]`); with `num_stages>=2` the prefetched read of `G[ko]` is emitted before the atomic to `G[ko]`, so every read sees the pre-atomic value (`[0,1,2,…,7]`). `G` itself ends `[0,101,102,…,108]` at every stage count — so **every atomic ran**; the read was reordered before the write, not dropped. It is single-threaded (`threads=1`), so this is a compile-time reordering, not a data race.

<details>
<summary>Emitted CUDA for <code>num_stages=2</code> — the prefetched read is placed before the atomic to the same element</summary>

```cuda
As[0] = G[0];                              // prologue prefetch of ko=0's read
As[((ko + 1) & 1)] = G[(ko + 1)];          // prefetch next iter's read of G[ko+1]  ── hoisted ABOVE ──
AtomicAdd((&(G[(ko + 1)])), 100);          // ── the write to that SAME element G[ko+1]
Out[ko] = As[(ko & 1)];
...
AtomicAdd((&(G[8])), 100);                 // epilogue
Out[7] = As[1];
```

The prefetch `As[(ko+1)&1] = G[ko+1]` reads `G[ko+1]` one line before `AtomicAdd(&G[ko+1], 100)` writes it. Because the pipeliner does not see the atomic as writing `G`, it treats the read as independent and hoists it into the prefetch stage.
</details>

<details>
<summary>Full stage table + control (init <code>G=[0..8]</code>, 2 runs each, sm_90/H100)</summary>

```
stages | Out                              | G_final                          | det   verdict
  1    | [0,101,102,103,104,105,106,107]  | [0,101,102,...,108]              | True  CORRECT (serial reference)
  2    | [0,  1,  2,  3,  4,  5,  6,  7]  | [0,101,102,...,108]              | True  WRONG (pre-atomic values)
  3    | [0,  1,  2,  3,  4,  5,  6,  7]  | [0,101,102,...,108]              | True  WRONG
  4    | [0,  1,  2,  3,  4,  5,  6,  7]  | [0,101,102,...,108]              | True  WRONG
```

- `num_stages=1` is the serial reference (software pipelining must preserve it exactly).
- `G_final` is `[0,101,…,108]` at *every* stage count → all 8 atomics executed; the wrong `Out` is a reordered read, not a lost write.
- `det=True` across 2 runs (and on re-run) at every stage count → deterministic, not a race.
</details>

### Reproducible example code

```python
import tilelang, tilelang.language as T, torch

N = 8
def make(stages):
    @tilelang.jit
    def k(G: T.Tensor((N + 1,), "int32"), Out: T.Tensor((N,), "int32")):
        with T.Kernel(1, threads=1):
            As = T.alloc_shared((1,), "int32")
            for ko in T.Pipelined(N, num_stages=stages):
                T.atomic_add(G[ko + 1], 100)   # write G[ko+1]
                As[0] = G[ko]                  # read G[ko]  (RAW on the previous iter's atomic to G[ko])
                Out[ko] = As[0]
    return k

def run(stages):
    G = torch.arange(0, N + 1, dtype=torch.int32, device="cuda")  # [0,1,2,...,8]
    Out = torch.zeros(N, dtype=torch.int32, device="cuda")
    make(stages)(G, Out)
    return Out.tolist()

print(run(1))  # -> [0, 101, 102, 103, 104, 105, 106, 107]   correct (serial semantics)
print(run(2))  # -> [0,   1,   2,   3,   4,   5,   6,   7]   WRONG: reads see pre-atomic G[ko]=ko
```

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic.
```

### Expected behavior

Software pipelining is a pure reordering optimization: the result at `num_stages>=2` must equal the `num_stages=1` result. The compiler should account for the atomic's write to `G` when it builds the stage dependency graph — so the read of `G[ko]` is kept after the write to `G[ko]` (serialized / not hoisted across the atomic) — or, if the atomic-carried dependency cannot be pipelined, decline to pipeline the loop. It should never silently reorder into a wrong answer. (A plain `BufferStore` write to the same buffer is recorded as a write and correctly blocks the reorder; the atomic should be treated the same way.)

### Additional context

**Root cause.** The pipeline dependency analysis records an atomic's buffer access as **read-only**, so the loop-carried RAW edge through the atomic is never built and the dependent read is free to hoist.

<details>
<summary>Where the write is dropped (pinned source)</summary>

- In [`BufferRegionCollector::VisitExpr_(CallNode)`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/pipeline/access_analysis.cc#L164-L173), the [`tvm_access_ptr` branch](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/pipeline/access_analysis.cc#L164-L173) unconditionally does `reads_.push_back(...)` and never `writes_.push_back(...)`; `op->args[4]` (the read/write mask) is never inspected. The [`address_of` branch](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/pipeline/access_analysis.cc#L149-L163) is likewise read-only. Contrast [`VisitStmt_(BufferStoreNode)`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/pipeline/access_analysis.cc#L104-L113), which does `writes_.push_back` — which is why a plain store to the same buffer correctly blocks the reorder.
- An `atomic_add` destination reaches the collector as `access_ptr(dst, "rw")` (`tvm_access_ptr` with `rw_mask=3`). The `atomic_*_elem_op` builtin has no tile-op builder, so `ParseOperator` returns nothing and the generic recursion (the `else` at [access_analysis.cc:181-182](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/pipeline/access_analysis.cc#L181-L183)) descends into the inner `tvm_access_ptr` — where the write bit is dropped.
- [`CollectPipelineWriteBuffers`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/pipeline/body_analysis.h#L36-L45) builds `pipeline_write_buffers` from `collector.GetWrites()`, so the atomic's buffer is absent from the write set → the stage RAW/WAR dependency graph never gets the edge → the read is classed independent and hoisted into the prefetch stage.
</details>

**Suggested fix.** In `BufferRegionCollector`, inspect the read/write mask `args[4]` on the `tvm_access_ptr` branch and `writes_.push_back(...)` when the write bit (`mask & 2`) is set; treat `address_of` args conservatively as writes; and/or recognize an `atomic_*_elem_op` destination as a write. Any of these restores the missing RAW edge so the pipeliner serializes correctly. Not built.

**Provenance.** The read-only classification of `tvm_access_ptr` in the pipeline access collector predates the current file: it was present in the pre-refactor monolithic `src/transform/inject_pipeline.cc` (`AddBuffer` on `tvm_access_ptr` with no mask check), and PR [#2245](https://github.com/tile-ai/tilelang/pull/2245) *"[Pipeline] Refactor software pipeline transforms"* (LeiWang1999, merged 2026-05-26) only **relocated** it into `src/transform/pipeline/access_analysis.cc`. So this is **not a regression** — it has been present since the pipeline access analysis was first written. Still live on tip-of-`main`: the collector was later inlined into `src/transform/pipeline_planning.cc`, but it still records `tvm_access_ptr` as read-only with no rw-mask check. Reproduced this session on sm_90 (H100); the pass is target-independent TIR (no `__CUDA_ARCH`/`sm_XX` fork around it), so other arches are INFERRED to behave identically (not run on sm_70/75/80/86).

**Related.** I searched the open and closed tracker (pipeline+atomic, num_stages+atomic, atomic RAW/stale-read) and found no existing report of this defect. The underlying pattern — an atomic destination classified read-only by an access analyzer, so a dependent read is reordered past it — also affects `IfStmtBinding` (`src/transform/if_stmt_binding.cc`) on the `if`-without-`else` bind-replay path, which I am reporting separately; the fix here (rw-mask on `tvm_access_ptr`) is local to the pipeline collector and does not cover that pass. #1366 (AtomicAdd auto-vectorization ignoring `BufferLayout`) and #2548 (Hopper warp-specialization ICHECK abort on `T.Pipelined`-inside-`while`) are different passes and symptoms.

**Comment by Hughshine** [comment]
The same reorder reproduces for the whole scalar-atomic write family, not just `atomic_add` — worth noting so a fix isn't scoped to one primitive. In each case a loop-carried read `Out[ko] = G[ko]` (staged through shared so it's a prefetch candidate) is hoisted ahead of an atomic write to `G[ko+1]`. `threads=1` (compile-time reorder, not a race), deterministic:

| primitive | access intent | ns<=1 (correct) | ns>=2 (wrong) |
|---|---|---|---|
| `atomic_add(G[ko+1],100)` | rw | `[0,101,...,107]` | `[0,1,...,7]` |
| `atomic_max(G[ko+1],1000)` | rw | `[0,1000,...,1000]` | `[0,0,...,0]` |
| `atomic_store(G[ko+1],900)` | **w (pure write)** | `[0,900,...,900]` | `[0,0,...,0]` |

At ns>=2 every atomic still executes (`G` ends up fully written) — only the read moved. `atomic_store` is the sharpest: a pure store is treated as read-only, so a value that was only ever written looks read-only to the scheduler.

They all lower through the same `access_ptr(dst, ...)` intrinsic, and the [`tvm_access_ptr` branch](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/pipeline/access_analysis.cc#L164-L172) does `reads_.push_back(...)` unconditionally without inspecting the access mask (`args[4]`). Reading that mask and pushing to `writes_` when the write bit is set would cover add/max/min/store in one place, rather than special-casing `atomic_add`.

(A bare global read isn't pipelined so it doesn't reorder — the read has to be a prefetch candidate. TileLang 0.1.11 @ `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`.)
